pkg/controller/datavolume/snapshot-clone-controller.go

```diff
@@ -0,0 +1,488 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+
+	cdiv1 "kubevirt.io/containerized-data-importer-api/pkg/apis/core/v1beta1"
+	"kubevirt.io/containerized-data-importer/pkg/common"
+	"kubevirt.io/containerized-data-importer/pkg/util"
+
+	cc "kubevirt.io/containerized-data-importer/pkg/controller/common"
+	featuregates "kubevirt.io/containerized-data-importer/pkg/feature-gates"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// SnapshotCloneReconciler members
+type SnapshotCloneReconciler struct {
+	CloneReconcilerBase
+}
+
+// NewSnapshotCloneController creates a new instance of the datavolume clone controller
+func NewSnapshotCloneController(
+	ctx context.Context,
+	mgr manager.Manager,
+	log logr.Logger,
+	clonerImage string,
+	importerImage string,
+	pullPolicy string,
+	tokenPublicKey *rsa.PublicKey,
+	tokenPrivateKey *rsa.PrivateKey,
+	installerLabels map[string]string,
+) (controller.Controller, error) {
+	client := mgr.GetClient()
+	reconciler := &SnapshotCloneReconciler{
+		CloneReconcilerBase: CloneReconcilerBase{
+			ReconcilerBase: ReconcilerBase{
+				client:          client,
+				scheme:          mgr.GetScheme(),
+				log:             log.WithName(snapshotCloneControllerName),
+				featureGates:    featuregates.NewFeatureGates(client),
+				recorder:        mgr.GetEventRecorderFor(snapshotCloneControllerName),
+				installerLabels: installerLabels,
+			},
+			clonerImage:    clonerImage,
+			importerImage:  importerImage,
+			pullPolicy:     pullPolicy,
+			tokenValidator: cc.NewCloneTokenValidator(common.CloneTokenIssuer, tokenPublicKey),
+			// for long term tokens to handle cross namespace dumb clones
+			tokenGenerator: newLongTermCloneTokenGenerator(tokenPrivateKey),
+		},
+	}
+	reconciler.Reconciler = reconciler
+
+	dataVolumeCloneController, err := controller.New(snapshotCloneControllerName, mgr, controller.Options{
+		Reconciler: reconciler,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if err := addDataVolumeSnapshotCloneControllerWatches(mgr, dataVolumeCloneController); err != nil {
+		return nil, err
+	}
+
+	return dataVolumeCloneController, nil
+}
+
+func addDataVolumeSnapshotCloneControllerWatches(mgr manager.Manager, datavolumeController controller.Controller) error {
+	if err := addDataVolumeControllerCommonWatches(mgr, datavolumeController, dataVolumeSnapshotClone); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Reconcile loop for the clone data volumes
+func (r SnapshotCloneReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithValues("DataVolume", req.NamespacedName)
+	return r.updateStatus(r.sync(log, req))
+}
+
+func (r SnapshotCloneReconciler) prepare(syncRes *dataVolumeSyncResult) error {
```

Comment:
looks similar to `PvcCloneReconciler.prepare()` so why not move to `CloneReconcilerBase`?

Reply:
yeah they happen to be the same but prefer to keep a specific one for each controller
not at peace with the claim that preparation will be the same for all clone controllers